### PR TITLE
Abstract Syntax List

### LINKED ISSUE #6: Make SkipForward and SkipBackword Constant Running Time
Currently the implementations of both `[` and `]` require a lookup through the code to find the matching brace each time. This should be avoided by doing some analysis up front (when the code is first loaded). This is also the reason for `new` #5 returning a `Result` because the code could be invalid when there are unmatched brackets.

### PR DESCRIPTION
**Pull request by nixpulvis**
Don't rely on the raw source code for the operation of the program directly. See #13 

Fixes #6

### CODE REVIEW

**Comment by nixpulvis** [comment]
Also fixes the `complex` smoke test.

File: src/program.rs [hunk 1]
```diff
@@ -15,23 +16,58 @@ impl Program {
...
         try!(file.read_to_string(&mut source));
-        Ok(Program::from_source(source))
+        Ok(Program::parse(&source))
     }
 
-    pub fn from_source<C: Into<String>>(source: C) -> Program {
-        Program {
-            source: source.into(),
+    pub fn parse(source: &str) -> Program {
+        let bracket_map = Program::bracket_map(source);
```
Comment (nixpulvis) [inline src/program.rs:21]: I'd love a better implementation than this.

**Comment by offlinemark** [comment]
drafted up a pseudocode-1 pass parse on paper, will code up later. pretty simple, just moves some of the stack operations/bracket index resolutions seen in bracket_map into the main parse loop

**Comment by nixpulvis** [comment]
Yea I was thinking something _like_ that would work.

**Comment by nixpulvis** [comment]
@mossberg feel free to commit directly to this branch.

**Comment by nixpulvis** [comment]
@mossberg I'm merging this cause I need it. Just make the patch on master.

**Pull request merged**